File: ruby/lesson05/lesson05.rb
# This code was created by Jeff Molofee '99 
# Conversion to Ruby by Manolo Padron Martinez (manolopm@cip.es)

require "opengl"
require "glut"


# A general OpenGL initialization function.  Sets all of the initial parameters

def InitGL(width, height) # We call this right after our OpenGL window 
                          # is created.

  GL.ClearColor(0.0, 0.0, 0.0, 0.0) # This Will Clear The Background 
                                    # Color To Black
  GL.ClearDepth(1.0)                # Enables Clearing Of The Depth Buffer
  GL.DepthFunc(GL::LESS)            # The Type Of Depth Test To Do
  GL.Enable(GL::DEPTH_TEST)         # Enables Depth Testing
  GL.ShadeModel(GL::SMOOTH)         # Enables Smooth Color Shading
  GL.MatrixMode(GL::PROJECTION)
  GL.LoadIdentity()                 # Reset The Projection Matrix
  GLU.Perspective(45.0,Float(width)/Float(height),0.1,100.0) # Calculate The Aspect Ratio 
                                               # Of The Window
  GL.MatrixMode(GL::MODELVIEW)
end

# The function called when our window is resized (which shouldn't happen, 
# because we're fullscreen) 
ReSizeGLScene = Proc.new {|width, height|
  if (height==0) # Prevent A Divide By Zero If The Window Is Too Small
    height=1
  end
  GL.Viewport(0,0,width,height) # Reset The Current Viewport And
                                # Perspective Transformation
  GL.MatrixMode(GL::PROJECTION)
  GL.LoadIdentity()
  GLU.Perspective(45.0,Float(width)/Float(height),0.1,100.0)
  GL.MatrixMode(GL::MODELVIEW)
}

# The main drawing function. 
DrawGLScene = Proc.new {
  GL.Clear(GL::COLOR_BUFFER_BIT | GL::DEPTH_BUFFER_BIT) # Clear The Screen And
                                          # The Depth Buffer
  GL.LoadIdentity()                       # Reset The View
  GL.Translate(-1.5, 0.0, -6.0)           # Move Left 1.5 Units And Into The 
                                          # Screen 6.0
  GL.Rotate($rtri,0.0,1.0,0.0)            # Rotate the triangle on the Y Axis

  # draw a triangle (in smooth coloring mode)
  GL.Begin(GL::POLYGON)                   # start drawing a polygon

  # front face of pyramid
  GL.Color3f(  1.0, 0.0, 0.0)             # Set The Color To Red
  GL.Vertex3f( 0.0, 1.0, 0.0)             # Top
  GL.Color3f(  0.0, 1.0, 0.0)             # Set The Color To Green
  GL.Vertex3f( 1.0,-1.0, 0.0)             # Bottom Right
  GL.Color3f(  0.0, 0.0, 1.0)             # Set The Color To Blue
  GL.Vertex3f(-1.0,-1.0, 0.0)             # Bottom Left  

  # right face of pyramid
  GL.Color3f(  1.0, 0.0, 0.0)             # Red
  GL.Vertex3f( 0.0, 1.0, 0.0)             # Top of triangle (Right)
  GL.Color3f(  0.0, 0.0, 1.0)             # Blue
  GL.Vertex3f( 1.0,-1.0, 1.0)             # Left of triangle (Right)
  GL.Color3f(  1.0, 0.0, 0.0)             # Green
  GL.Vertex3f( 1.0,-1.0,-1.0)             # Right of triangle (Right)

  # back face of pyramid
  GL.Color3f(  1.0, 0.0, 0.0)             # Red
  GL.Vertex3f( 0.0, 1.0, 0.0)             # Top of triangle (Back)
  GL.Color3f(  0.0, 1.0, 0.0)             # Green
  GL.Vertex3f( 1.0,-1.0,-1.0)             # Left of triangle (Back)
  GL.Color3f(  0.0, 0.0, 1.0)             # Blue
  GL.Vertex3f(-1.0,-1.0,-1.0)             # Right of triangle (Back

  # left face of pyramid
  GL.Color3f(  1.0, 0.0, 0.0)             # Red
  GL.Vertex3f( 0.0, 1.0, 0.0)             # Top of triangle (Left)
  GL.Color3f(  0.0, 0.0, 1.0)             # Blue
  GL.Vertex3f(-1.0,-1.0,-1.0)             # Left of triangle (Left)
  GL.Color3f(  1.0, 0.0, 0.0)             # Green
  GL.Vertex3f(-1.0,-1.0, 1.0)             # Right of triangle (Left)

  GL.End()                                # Done drawing the pyramid

  GL.LoadIdentity()                       # make sure we're no longer rotated.
  GL.Translate(1.5,0.0,-7.0)              # Move Right 3 Units, and back into 
                                          # the screen 7.0
  GL.Rotate($rquad,1.0,1.0,1.0)           # Rotate the quad on the X Axis
  # draw a cube (6 quadrilateral)
  GL.Begin(GL::QUADS)                     # start drawing the cube

  # top of cube
  GL.Color3f(0.0,1.0,0.0)               # Set The Color To Blue
  GL.Vertex3f( 1.0, 1.0,-1.0)             # Top Right Of The Quad (Top)
  GL.Vertex3f(-1.0, 1.0,-1.0)             # Top Left Of The Quad (Top)
  GL.Vertex3f(-1.0, 1.0, 1.0)             # Bottom Left Of The Quad (Top)
  GL.Vertex3f( 1.0, 1.0, 1.0)             # Bottom Right Of The Quad (Top)

  # bottom of cube
  GL.Color3f(1.0,0.5,0.0)             	# Set The Color To Orange
  GL.Vertex3f( 1.0,-1.0, 1.0)             # Top Right Of The Quad (Bottom)
  GL.Vertex3f(-1.0,-1.0, 1.0)             # Top Left Of The Quad (Bottom)
  GL.Vertex3f(-1.0,-1.0,-1.0)             # Bottom Left Of The Quad (Bottom)
  GL.Vertex3f( 1.0,-1.0,-1.0)             # Bottom Right Of The Quad (Bottom)

  # front of cube
  GL.Color3f(1.0,0.0,0.0)             	# Set The Color To Red
  GL.Vertex3f( 1.0, 1.0, 1.0)             # Top Right Of The Quad (Front)
  GL.Vertex3f(-1.0, 1.0, 1.0)             # Top Left Of The Quad (Front)
  GL.Vertex3f(-1.0,-1.0, 1.0)             # Bottom Left Of The Quad (Front)
  GL.Vertex3f( 1.0,-1.0, 1.0)             # Bottom Right Of The Quad (Front)

  # back of cube.
  GL.Color3f(1.0,1.0,0.0)             	# Set The Color To Yellow
  GL.Vertex3f( 1.0,-1.0,-1.0)             # Top Right Of The Quad (Back)
  GL.Vertex3f(-1.0,-1.0,-1.0)             # Top Left Of The Quad (Back)
  GL.Vertex3f(-1.0, 1.0,-1.0)             # Bottom Left Of The Quad (Back)
  GL.Vertex3f( 1.0, 1.0,-1.0)             # Bottom Right Of The Quad (Back)

  # left of cube
  GL.Color3f(0.0,0.0,1.0)             	# Blue
  GL.Vertex3f(-1.0, 1.0, 1.0)             # Top Right Of The Quad (Left)
  GL.Vertex3f(-1.0, 1.0,-1.0)             # Top Left Of The Quad (Left)
  GL.Vertex3f(-1.0,-1.0,-1.0)             # Bottom Left Of The Quad (Left)
  GL.Vertex3f(-1.0,-1.0, 1.0)             # Bottom Right Of The Quad (Left)

  # Right of cube
  GL.Color3f(1.0,0.0,1.0)             	# Set The Color To Violet
  GL.Vertex3f( 1.0, 1.0,-1.0);	          # Top Right Of The Quad (Right)
  GL.Vertex3f( 1.0, 1.0, 1.0)             # Top Left Of The Quad (Right)
  GL.Vertex3f( 1.0,-1.0, 1.0)             # Bottom Left Of The Quad (Right)
  GL.Vertex3f( 1.0,-1.0,-1.0)             # Bottom Right Of The Quad (Right)
  GL.End();                               # done with the polygon

  $rtri=$rtri+15.0                        # Increase the rotation variable for
                                          # the Triangle
  $rquad=$rquad-15.0                      # Decrease the rotation variable for 
                                          # the Quad
  # we need to swap the buffer to display our drawing.
  GLUT.SwapBuffers();
}



# The function called whenever a key is pressed.
keyPressed = Proc.new {|key, x, y| 

  # If escape is pressed, kill everything. 
  if (key == 27) 
    # shut down our window 
    GLUT.DestroyWindow($window)
    # exit the program...normal termination.
    exit(0)                   
  end
}

# Rotation angle for the triangle.
$rtri=0.0

#Rotation angle for the quadrilateral.
$rquad=0.0

#Initialize GLUT state - glut will take any command line arguments that pertain
# to it or X Windows - look at its documentation at 
# http://reality.sgi.com/mjk/spec3/spec3.html 
GLUT.Init

#Select type of Display mode:   
# Double buffer 
# RGBA color
# Alpha components supported 
# Depth buffer 
GLUT.InitDisplayMode(GLUT::RGBA|GLUT::DOUBLE|GLUT::ALPHA|GLUT::DEPTH)

# get a 640x480 window
GLUT.InitWindowSize(640,480)

# the window starts at the upper left corner of the screen
GLUT.InitWindowPosition(0,0)

# Open a window
$window=GLUT.CreateWindow("Jeff Molofee's GL Code Tutorial ... NeHe '99")

# Register the function to do all our OpenGL drawing.
GLUT.DisplayFunc(DrawGLScene)

# Go fullscreen. This is as soon as possible.
GLUT.FullScreen()

# Even if there are no events, redraw our gl scene.
GLUT.IdleFunc(DrawGLScene)

# Register the function called when our window is resized.
GLUT.ReshapeFunc(ReSizeGLScene)

# Register the function called when the keyboard is pressed.
GLUT.KeyboardFunc(keyPressed)

# Initialize our window.
InitGL(640, 480)

# Start Event Processing Engine
GLUT.MainLoop()

File: perl/lesson05/lesson05.pl
#! /usr/bin/perl -w
#
# This code was created by Jeff Molofee '99 
# Ported to Perl by Cora Hussey '02
#
# If this was helpful, let me know.
# Visit me at www.cs.ucla.edu/~chussey
#
# Requires OpenGL module in addition to OpenGL
#   (visit www.cpan.org or "perl -MCPAN -e 'install OpenGL'")
#

use OpenGL qw(:all);    # Use the OpenGL module
use strict;             # Use strict typechecking

# ASCII constant for the escape key
use constant ESCAPE => 27;

# Global variable for our window
my $window;

# Rotation variables for triangle and quad
my $rtri  = 0.0;
my $rquad = 0.0;

# A general GL initialization function 
# Called right after our OpenGL window is created
# Sets all of the initial parameters
sub InitGL {              

    # Shift the width and height off of @_, in that order
    my ($width, $height) = @_;

    # Set the background "clearing color" to black
    glClearColor(0.0, 0.0, 0.0, 0.0);

    # Enables clearing of the Depth buffer 
    glClearDepth(1.0);                    

    # The type of depth test to do
    glDepthFunc(GL_LESS);         

    # Enables depth testing with that type
    glEnable(GL_DEPTH_TEST);              
    
    # Enables smooth color shading
    glShadeModel(GL_SMOOTH);      

    # Reset the projection matrix
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity;

    # Calculate the aspect ratio of the Window
    gluPerspective(45.0, $width/$height, 0.1, 100.0);

    # Reset the modelview matrix
    glMatrixMode(GL_MODELVIEW);
}


# The function called when our window is resized 
# This shouldn't happen, because we're fullscreen
sub ReSizeGLScene {

    # Shift width and height off of @_, in that order
    my ($width, $height) = @_;

    # Prevent divide by zero error if window is too small
    if ($height == 0) { $height = 1; }

    # Reset the current viewport and perspective transformation
    glViewport(0, 0, $width, $height);              

    # Re-initialize the window (same lines from InitGL)
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity;
    gluPerspective(45.0, $width/$height, 0.1, 100.0);
    glMatrixMode(GL_MODELVIEW);
}

# The main drawing function.
sub DrawGLScene {

    # Clear the screen and the depth buffer
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);  

    # Reset the view
    glLoadIdentity;

    # Move to the left 1.5 units and into the screen 6.0 units
    glTranslatef(-1.5, 0.0, -6.0); 
    
    # Rotate the triangle on the y axis 
    # Order is (degrees, x, y, z), set to 1.0 your rotation axis
    glRotatef($rtri, 0.0, 1.0, 0.0); 
    
    # -- Draw a pyramid (in smooth coloring mode) --

    # Begin drawing a polygon
    # Note that each face (of 4) must be shaded
    glBegin(GL_POLYGON);

    # Front face of pyramid
    glColor3f(1.0, 0.0, 0.0);       # Set the color to red
    glVertex3f(0.0, 1.0, 0.0);      # Top of triangle (front of pyramid)
    glColor3f(0.0, 1.0, 0.0);       # Set the color to green
    glVertex3f(-1.0, -1.0, 1.0);    # Left of triangle (front of pyramid)
    glColor3f(0.0, 0.0, 1.0);       # Set the color to blue
    glVertex3f(1.0, -1.0, 1.0);     # Right of traingle (front of pyramid)    

    # Right face of pyramid
    glColor3f(1.0, 0.0, 0.0);       # Red color
    glVertex3f(0.0, 1.0, 0.0);      # Top of triangle (right of pyramid)
    glColor3f(0.0, 0.0, 1.0);       # Blue color
    glVertex3f(1.0, -1.0, 1.0);     # Left of triangle (right of pyramid)
    glColor3f(0.0, 1.0, 0.0);       # Green color
    glVertex3f(1.0, -1.0, -1.0);    # Right of triangle (right of pyramid)

    # Back face of pyramid
    glColor3f(1.0, 0.0, 0.0);       # Red color
    glVertex3f(0.0, 1.0, 0.0);      # Top of triangle (back of pyramid)
    glColor3f(0.0, 1.0, 0.0);       # Green color
    glVertex3f(1.0, -1.0, -1.0);    # Left of triangle (back of pyramid)
    glColor3f(0.0, 0.0, 1.0);       # Blue color
    glVertex3f(-1.0, -1.0, -1.0);   # Right of triangle (back of pyramid)

    # Left face of pyramid.
    glColor3f(1.0, 0.0, 0.0);       # Red color
    glVertex3f(0.0, 1.0, 0.0);      # Top of triangle (left of pyramid)
    glColor3f(0.0, 0.0, 1.0);       # Blue color
    glVertex3f(-1.0, -1.0, -1.0);   # Left of triangle (left of pyramid)
    glColor3f(0.0, 1.0, 0.0);       # Green color
    glVertex3f(-1.0, -1.0, 1.0);    # Right of triangle (left of pyramid)

    # Done with the polygon
    glEnd;

    # Reset the screen
    glLoadIdentity;

    # Move to the right 1.5 units and into the screen 6.0 units
    glTranslatef(1.5, 0.0, -6.0); 

    # Rotate the quad on the x y and z axis
    glRotatef($rquad, 1.0, 1.0, 1.0);
        
    # -- Draw a cube (6 quadrilaterals) --

    # Set color to a blue color
    glColor3f(0.5, 0.5, 1.0);

    # Begin drawing the cube
    glBegin(GL_QUADS);

    # Top face of cube
    glColor3f(0.0, 1.0, 0.0);       # Set the color to Blue
    glVertex3f(1.0, 1.0, -1.0);     # Top right vertex (Top of cube)
    glVertex3f(-1.0, 1.0, -1.0);    # Top left vertex (Top of cube)
    glVertex3f(-1.0, 1.0, 1.0);     # Bottom left vertex (Top of cube)
    glVertex3f(1.0, 1.0, 1.0);      # Bottom right vertex (Top of cube)

    # Bottom face of cube
    glColor3f(1.0, 0.5, 0.0);       # Set the color to Orange
    glVertex3f(1.0, -1.0, 1.0);     # Top right vertex (Bottom of cube)
    glVertex3f(-1.0, -1.0, 1.0);    # Top left vertex (Bottom of cube)
    glVertex3f(-1.0, -1.0, -1.0);   # Bottom left vertex (Bottom of cube)
    glVertex3f( 1.0, -1.0, -1.0);   # Bottom right vertex (Bottom of cube)

    # Front of cube
    glColor3f(1.0, 0.0, 0.0);       # Set the color to Red
    glVertex3f(1.0, 1.0, 1.0);      # Top right vertex (Front)
    glVertex3f(-1.0, 1.0, 1.0);     # Top left vertex (Front)
    glVertex3f(-1.0, -1.0, 1.0);    # Bottom left vertex (Front)
    glVertex3f(1.0, -1.0, 1.0);     # Bottom right vertex (Front)

    # Back of cube.
    glColor3f(1.0, 1.0, 0.0);       # Set the color to Yellow
    glVertex3f(1.0, -1.0, -1.0);    # Top right vertex (Back)
    glVertex3f(-1.0, -1.0, -1.0);   # Top left vertex (Back)
    glVertex3f(-1.0, 1.0, -1.0);    # Bottom left vertex (Back)
    glVertex3f(1.0, 1.0, -1.0);     # Bottom right vertex (Back)

    # Left of cube
    glColor3f(0.0, 0.0, 1.0);       # Blue
    glVertex3f(-1.0, 1.0, 1.0);     # Top right vertex (Left)
    glVertex3f(-1.0, 1.0, -1.0);    # Top left vertex (Left)
    glVertex3f(-1.0, -1.0, -1.0);   # Bottom left vertex (Left)
    glVertex3f(-1.0, -1.0, 1.0);    # Bottom vertex (Left)

    # Right of cube
    glColor3f(1.0, 0.0, 1.0);       # Violet
    glVertex3f(1.0, 1.0, -1.0);     # Top right vertex (Right)
    glVertex3f(1.0, 1.0, 1.0);      # Top left vertex (Right)
    glVertex3f(1.0, -1.0, 1.0);     # Bottom left vertex (Right)
    glVertex3f(1.0, -1.0, -1.0);    # Bottom right vertex (Right)

    # End drawing the cube
    glEnd;                

    # Increase rotation angle for triangle, decrease for quad
    $rtri  = $rtri  + 15.0; 
    $rquad = $rquad - 15.0; 

    # Since this is double buffered, swap the buffers.
    # This will display what just got drawn.
    glutSwapBuffers;
}

# The function called whenever a key is pressed. 
sub keyPressed {

    # Shift the unsigned char key, and the x,y placement off @_, in
    # that order.
    my ($key, $x, $y) = @_;
    
    # Avoid thrashing this procedure
    # Note standard Perl does not support usleep
    # For finer resolution sleep than seconds, try:
    #    'select undef, undef, undef, 0.1;'
    # to sleep for (at least) 0.1 seconds
    sleep(100);

    # If f key pressed, undo fullscreen and resize to 640x480
    if ($key == ord('f')) {

        # Use reshape window, which undoes fullscreen
        glutReshapeWindow(640, 480);
    }

    # If escape is pressed, kill everything.
    if ($key == ESCAPE) 
    { 
        # Shut down our window 
        glutDestroyWindow($window); 
        
        # Exit the program...normal termination.
        exit(0);                   
    }
}

# --- Main program ---

# Initialize GLUT state
glutInit;  

# Select type of Display mode:   
# Double buffer 
# RGB color (Also try GLUT_RGBA)
# Alpha components removed (try GLUT_ALPHA) 
# Depth buffer */  
glutInitDisplayMode(GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH);  

# Get a 640 x 480 window
glutInitWindowSize(640, 480);  

# The window starts at the upper left corner of the screen
glutInitWindowPosition(0, 0);  

# Open the window  
$window = glutCreateWindow("Jeff Molofee's GL Code Tutorial ... NeHe '99");  

# Register the function to do all our OpenGL drawing.
glutDisplayFunc(\&DrawGLScene);  

# Go fullscreen.  This is as soon as possible. 
glutFullScreen;

# Even if there are no events, redraw our gl scene.
glutIdleFunc(\&DrawGLScene);

# Register the function called when our window is resized. 
glutReshapeFunc(\&ReSizeGLScene);

# Register the function called when the keyboard is pressed.
glutKeyboardFunc(\&keyPressed);

# Initialize our window.
InitGL(640, 480);
  
# Start Event Processing Engine
glutMainLoop;  

return 1;


